game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/BattleCalculator.java

```diff
@@ -25,6 +26,8 @@
 
 class BattleCalculator implements IBattleCalculator {
   @Nonnull private final GameData gameData;
+  // Use a single TuvCostsCalculator so its computations are cached.
+  private final TuvCostsCalculator tuvCalculator = new TuvCostsCalculator();
```

Comment:
Is this absolutely crucial for the performance improvements. I feel like in a lot of cases it would  be sufficient to have a field like this in the individual classes that require it, but I might be completely wrong here

Reply:
Yes, it's part of the major performance speed up. The BattleCalculator is re-used to perform different computations with the same overall GameData state.